### ISSUE TITLE
Slow figwheel hot reload

### ISSUE THREAD

**Issue by bruno-fleischer** [bhauman/figwheel-main#215]
Hi, I'm having issues trying to improve the compiling times of figwheel when hotreloading.

I'm not sure if it's happening because of the size of the project or because of some wrong configuration. It's taking about 14 seconds to hot reload, and sometimes it gets stuck and takes longer than 60 seconds.

My current configuration in dev.cljs.edn is:

`^{:watch-dirs ["dev/cljs" "src/cljs" "src/cljc" "test/cljs"]
  :css-dirs ["resources/public/css"]
  :auto-testing true
  :open-url false
  :log-level :debug
  :client-log-level :warning
  :load-warninged-code true}
{:main finops-admin.dev
 :asset-path "/env/dev"
 :output-dir "resources/public/env/dev"
 :output-to "resources/public/env/dev.js"
 :optimizations :none
 :compiler-stats true
 :parallel-build true
 :recompile-dependents false
 :source-map-timestamp true
 :verbose false
 :preloads [day8.re-frame-10x.preload]
 :closure-defines {"re_frame.trace.trace_enabled_QMARK_" true
                   "day8.re_frame.tracing.trace_enabled_QMARK_" true}
 :foreign-libs [{:file "src/js/honeybadger-js/honeybadger.js"
                 :provides ["honeybadger-js"]}]
 :warnings {:fn-deprecated false}
 :aot-cache true}`

**Comment by jordan-yee** [comment]
I'm also facing long compile times. The initial build takes about 90 seconds, then the first time I save a file it takes about 60 seconds, and then it stabilizes at about 20 seconds after that.

I have a fairly big project, but I don't think that's the problem--it's not _that_ big as far as SPAs go (1.9MB advanced compile).

And I get the ~20 second baseline compilation when saving a file with zero dependencies (no require).

**Comment by bhauman** [comment]
Hey I'd love to wake this up.  Sorry for not getting back sooner. 

@bruno-fleischer it sounds like yours is happening on the front end?
On the front end you can use the `:client-log-level :finest` to help narrow it down for me.

If you suspect that its happening on the back end you can use the `jstack` tool.
H https://docs.oracle.com/javase/7/docs/technotes/tools/share/jstack.html

It could really help identify where things are hanging.  

Either way I'm happy to join you for a session so that we can see what's causing this.

**Comment by jordan-yee** [comment]
@bhauman I don't want to hijack @bruno-fleischer's issue, but I'm actively working on debugging a slow reload, and would be happy to provide additional information for my case, which may or may not  be the same problem.

**Comment by bhauman** [comment]
Heck yeah please do!

**Comment by jordan-yee** [comment]
About my project:
- It's a re-frame/reagent based SPA.
- I'm using figwheel-main via the scripting API, as described in the [Vim section](https://figwheel.org/docs/vim.html) of the documentation.
- Project size: 217 cljs files, 25 cljc files
- Compile times:
  - Initial compile time (after starting figwheel):
  `[Figwheel] Successfully compiled build dev to "target/public/cljs-out/<project>.js" in 97.459 seconds.`
  - Secondary compile time (after saving a file for the first time, or after modifying the require):
  `[Figwheel] Successfully compiled build dev to "target/public/cljs-out/<project>.js" in 76.872 seconds.`
    (I have no idea where this intermediate time comes from.)
  - Baseline compile time (after saving a file for the second+ time(s) without making changes to it).
  `[Figwheel] Successfully compiled build dev to "target/public/cljs-out/<project>.js" in 26.532 seconds.`
- Trimming out all client-side code and dependencies except the core.cljs entry point resulted in reload times of a couple of seconds, which seems to indicate that the problem has to do with my code, as opposed to system or configuration issues.

Here's my dev config:
```
^{:watch-dirs ["src/cljs" "src/cljc"]
  :css-dirs ["resources/public/css/compiled"]
  :open-url "http://localhost:3553"

  ;;; Logging ;;;
  ;:log-file "figwheel-main.log"
  ;:log-level :debug
  ;:client-log-level :finest
  ;:log-syntax-error-style :concise
  }
{:main <project>.core
 :output-to "target/public/cljs-out/<project>.js"
 :externs ["externs/stripe-v3.ext.js"]
 :infer-externs true
 :recompile-dependents false}
```

**Comment by jordan-yee** [comment]
UPDATE:
With the addition of WSL 2 to Windows, I converted my development setup to a primarily Linux-based one rather than Windows-based. With the same project I was testing above, I'm now seeing the following timings:

- Initial compilation: **130.034 seconds**
- Second compilation: **31.71 seconds**
- Third compilation: **10.872 seconds**
- Subsequent compilations: **8 to 10 seconds**

Obviously the goal is to have instantaneous recompile and reload of code changes, but these are a TREMENDOUS improvement over what I was getting when running everything in Windows.

tldr: There seem to be performance issues when running Figwheel-main in Windows.
